src/main/java/com/gmail/nossr50/util/blockmeta/chunkmeta/PrimitiveChunkStore.java

```diff
@@ -118,28 +133,43 @@ public void setFalse(int x, int y, int z) {
     }
 
     @Override
+    @Contract(pure = true)
     public boolean isEmpty() {
         int end = data.length();
+        // same as for(int i = end.length()-1; i >= 0; i--)
         while (end-- > 0) {
+            // If any bit is true, return false
             if (data.get(end) != 0) return false;
         }
+        // No bit is true
         return true;
     }
 
     @Override
     public void copyFrom(ChunkletStore otherStore) {
-        if (otherStore instanceof PrimitiveChunkStore) {
-            AtomicIntegerArray data = ((PrimitiveChunkStore) otherStore).data;
-            AtomicIntegerArray to = this.data = new AtomicIntegerArray(data.length());
-            int end = to.length();
+        if (otherStore instanceof PrimitiveChunkStore) { // Direct copy.
+            AtomicIntegerArray from = ((PrimitiveChunkStore) otherStore).data;
+            AtomicIntegerArray into = this.data = new AtomicIntegerArray(
+                    worldHeight * PAGE_SIZE
+            );
+            // To ensure that no cross-border
+            // Like:
+            // from: ---- ---- ---- ----
+            // into: ---- ---- ---- ---- ---- ---- ---- ----
+
+            // from: ---- ---- ---- ---- ---- ---- ---- ----
+            // into: ---- ---- ---- ----
+            int end = Math.min(from.length(), into.length());
+            // same as for(int i = end.length()-1; i >= 0; i--)
             while (end-- > 0) {
-                to.set(end, data.get(end));
+                into.set(end, from.get(end));
             }
         } else {
-            int wh = worldHeight;
+            // Copy values with legacy method.
```

Comment:
I'd prefer a comment explaining the process of this nested loop, I understand you didn't write it so I could let it slide.